packages/remark-math/inline.js

```diff
@@ -2,72 +2,124 @@ function locator (value, fromIndex) {
   return value.indexOf('$', fromIndex)
 }
 
-const ESCAPED_INLINE_MATH = /^\\\$/
-const INLINE_MATH = /^\$((?:\\\$|[^$])+)\$/
-const INLINE_MATH_DOUBLE = /^\$\$((?:\\\$|[^$])+)\$\$/
+const ESCAPED_INLINE_MATH = /^\\\$/ // starts with \$
+
+const NODE_TYPES = {
+  TEXT: 'text',
+  MATH_DISPLAY: 'math',
+  MATH_INLINE: 'inlineMath'
+}
+
+const defaultModes = {
+  mathInlineDouble: {
+    nodeType: NODE_TYPES.MATH_INLINE,
+    left: /\$\$/,
+    right: /\$\$/,
+    matchInclude: [/\\\$/, /[^$]/],
```

Comment:
I think I'd like to keep this as-is, if that's OK with you. What is an allowable internal match is related to what delimiters are used. I think that the fact that it's the same between the two expressions is somewhat a limitation of regular-expression-based parsing, and also the difficulties of deriving an inverse of a particular string. While this perhaps isn't ideal, it's the solution that seemed to be the best attainable ground to me, and it's (relatively) readable, and it works.

Reply:
I thought that too, but I'm not so sure if that's the reason - you just can't have a dollar sign by itself in KaTeX even if you aren't using dollar sign delimiters. E.g. the demo on https://katex.org/ doesn't use any delimiters to denote latex, it's just parsing the whole input field. But you still get an error if you put in a dollar sign.

Playing around with some examples (trying to see what happens if we replace the middle with just `.+`:

https://regexr.com/4uos3
https://regexr.com/4uos9

Ah I kinda see, I guess the purpose of this might just be to make it so that it ends the match at the first dollar sign if one exists, but in the case of double dollar signs that doesn't make sense.

Double dollar signs:

https://regexr.com/4uosl
https://regexr.com/4uoso

Hmm, given that, it does kinda makes sense to change the middle part for double dollar signs, but it's weird to me that they did not do that for remark-math's existing double dollar sign regex, and we also have not done that here. I think the way it's set up it might not do multi-line matching so we wouldn't have that weirdness of going across lines, but it definitely would not render LaTeX there unless you escape the dollar sign.

At a higher level though, I feel like this is something that a consumer of this API wouldn't necessarily want to think about. At the least having some defaults would be nice IMO. I also feel that it would be nice if we could pass in strings for the left and right delimiters rather than regexes, but that's a minor thing. Again, I think [KaTeX auto-render's API](https://katex.org/docs/autorender.html#api) is nice and simple with the right level of abstraction.